LG-3752: Handle failed uploads in document capture async upload

aduth opened the pull request:
**Why**: Currently, if encryption or async upload fails at the document capture step, we don't inform the user, so from their perspective it likely seem as if the upload was successful. With these changes, we'll show an error message indicating that there was an issue. This integrates into the existing FormSteps field error handling behavior which will prevent users from continuing until they change the value.

**Screenshot:**

![image](https://user-images.githubusercontent.com/1779930/110162737-20f2b500-7dbd-11eb-9be7-e96506ce72a3.png)

aduth commented on app/javascript/packages/document-capture/higher-order/with-background-encrypted-upload.jsx:
```diff
@@ -46,58 +52,66 @@ export async function encrypt(key, iv, value) {
-              body: encryptedValue,
-              headers: { 'Content-Type': 'application/octet-stream' },
-            }),
+  ({ onChange, onError, ...props }) => {
```
Recommend ["Hide Whitespace Changes" setting](https://user-images.githubusercontent.com/1779930/109327839-c4285500-7826-11eb-960d-d0fe88e69b25.png). It's just indentation.

zachmargolis replied:
Will our TS checks error if we construct this component and do not provide an `onError`?

aduth replied:
> Will our TS checks error if we construct this component and do not provide an `onError`?

The individual components should define the type themselves ([example](https://github.com/18F/identity-idp/blob/f70fa9894dc6fec44228b00b9c3e5827a61816f8/app/javascript/packages/document-capture/components/selfie-step.jsx#L23-L25)). Adding the type here was more a convenience to double-check the usage of `onChange` and `onError` within this implementation.

zachmargolis commented:
> * We should probably be logging something

Should we add a newrelic.addPageEvent?
   
> * File inputs don't emit a `change` event if the user selects a same file. Therefore, if the user "retries" with the same image, it won't actually try to upload. If this is problematic, we could consider to clear the value when the error is caught.

I think clearing the image would be bad, it can be a pain to find an image to upload or take a new one. Would it be clearer if we did the encryption + upload when users submit the form? Potentially a little slower, but makes for easier retries?

aduth commented:
> I think clearing the image would be bad, it can be a pain to find an image to upload or take a new one. Would it be clearer if we did the encryption + upload when users submit the form? Potentially a little slower, but makes for easier retries?

That's an interesting thought. We could potentially still attempt an initial encryption + upload, and if there was a failure, retry at the next submission, and _then_ if it's still a failure show a message, and continue to try reuploading in response to subsequent submissions.

anniehirshman-gsa commented:
From code review: if the result of an S3 upload is that something is going wrong on our end that is going to block the end user from proofing, we want to alert the user as soon as possible, so that they don't spend time taking more images if there's an issue they can't resolve. The ideal solution (for the user) would be to fall back on another method of submitting images so they can continue without disruption. But if not, then clearing the field would show that the image upload didn't work and allow them to try again with the same file (if uploading, otherwise take a new photo).

Attached a couple screens: one with a short field-level error and another with the combination of a field-level error and a more descriptive alert banner.  If 1 isn't clear enough on its own, maybe we add the alert banner (in all cases or after a re-try)?

![Image Capture - Back ID Added S3 Error - Option 1](https://user-images.githubusercontent.com/66273797/109533672-7c086d00-7a88-11eb-8860-8213659ac12e.png)

![Image Capture - Back ID Added S3 Error - Option 2](https://user-images.githubusercontent.com/66273797/109533669-7b6fd680-7a88-11eb-9e72-c9745f5b87f3.png)

aduth commented:
A few of us on Ada did a pilot code review on this pull request this morning.

Some topics covered:

- **At what point should we inform the user there was a problem, and how should the user be expected to move forward?**
   - As @anniehirshman-gsa mentioned above, there was some consensus to clearing and showing a message at the field rather than at submission, so that the user is informed as soon as possible, and so that they don't feel misled to believe that their image was selected when in-fact an error had occurred in uploading it.
      - Per @zachmargolis' point above, I do think it's a bit of unnecessary effort on the user's part to retake a photo on mobile, though I think I still prefer this over preserving the selected value as potentially implying a successful capture, or the user having to come back and inevitably deal with this issue if a subsequent re-upload during submission fails as well.
   - @solipet mentioned a possible alternative of falling back to the (current production) synchronous upload submission if async upload fails. While this means we'd have to continue maintaining this upload mechanism, it might prove to be a reliable escape hatch. I likely won't implement this just yet.
   - We could do one or more retries in the `catch` handling before finally showing the user a message, if we think there's a chance it could succeed on a redo. I likely won't implement this just yet.
- **What should be logged, and how?**
   - At least something to CloudWatch, as part of ongoing logging efforts to better understand where issues occur.
   - Ideally we break this down based on what sort of error happens. For example, we can differentiate if an error occurs during encryption vs. by uploading to S3, and in uploading errors, we may be able to pull some useful details from the response (e.g. HTTP status code?).
   - To @zachmargolis's comment above, sending events to both CloudWatch and NewRelic is fairly trivial using the `addPageAction` React context method.

zachmargolis commented on app/javascript/packages/document-capture/higher-order/with-background-encrypted-upload.jsx:
```diff
@@ -2,6 +2,21 @@ import { useContext } from 'react';
+ * An error representing a failure to complete encrypted upload of image.
+ */
+export class BackgroundEncryptedUploadError extends Error {
+  baseField = '';
+
+  /** @type {string[]} */
+  fields = [];
```
would it be worth our time to use the type system to restrict which fields these can be? `front` `back` `front_image_url` `back_image_url` etc etc etc via like a Record type?

aduth replied:
> would it be worth our time to use the type system to restrict which fields these can be? `front` `back` `front_image_url` `back_image_url` etc etc etc via like a Record type?

We could. As implemented currently, the `withBackgroundEncryptedUpload` doesn't really know anything about the specific fields it's being applied to, so in that sense a generic `string[]` makes sense. It'd be nice if we could be a bit more strict here at least as far as knowing that it'd be one "base" key, then two additional keys `${key}_image_iv` and `${key}_image_url`. Not sure if this is something we can really add type-checking for, though.

zachmargolis reviewed (approved):
LGTM

aduth commented:
Updates:

- "Clearing" the value didn't really have much of an advantage in addressing the problem of the browser not emitting a `change` event for existing values, since we don't really control the input so far as keeping it in sync with the values we're keeping for the form. Long story short, we have to dismiss the browser input field's value every time it changes, which helps ensure we always emit a `change` event, whether or not we clear the value on an error. (934414a)
- There was some complexity around dealing with the scenario where a user clicks "Submit" before the files have finished uploading to S3, because in those cases the form has been removed from the page by the time the upload resolves. This means we have to have a bit more awareness into these specific types of errors when showing the "Review" step. An alternative here could be to prevent submission until uploads finish, but this was something we'd decided earlier would not be as nice a user experience. (b64ad8d)
- Based on @anniehirshman-gsa's mockups above, I've updated the messaging. (d6b2583)
- I've added some additional logging to cover the case that an error is thrown during encryption, like we've seen happen before in #4705. We already were [logging errors](https://github.com/18F/identity-idp/blob/cd9df4af34fa4d27b28ac7deaaba7d28de617925/app/javascript/packages/document-capture/higher-order/with-background-encrypted-upload.jsx#L78-L85) for unsuccessful uploads with a trace ID of the request. (e834154)

aduth commented on spec/javascripts/packages/document-capture/components/document-capture-spec.jsx:
```diff
@@ -475,4 +475,90 @@ describe('document-capture/components/document-capture', () => {
+    });
+
+    context('success', () => {
+      it('calls to upload once pending values resolve', async () => {
+        submit();
+        expect(upload).not.to.have.been.called();
+        completeUploadAsSuccess();
+        await new Promise((resolve) => upload.callsFake(resolve));
+      });
```
Re: Failing build, I think there's a race condition happening somewhere in here.

aduth commented on app/javascript/packages/document-capture/components/file-input.jsx:
```diff
@@ -132,6 +132,17 @@ const FileInput = forwardRef((props, ref) => {
+      // can only be set by a user, and not programmatically."
+      //
+      // See: https://reactjs.org/docs/uncontrolled-components.html#the-file-input-tag
+      event.target.value = '';
```
In Internet Explorer, this appears to emit its own separate `change` event, thus clearing what what the form considers to be the current value of the field after making a selection.

aduth commented:
> In Internet Explorer, this appears to emit its own separate `change` event, thus clearing what what the form considers to be the current value of the field after making a selection.

It's a bit more indirect than I'd prefer, but c1ace23 should handle this. I gave it a pretty thorough round of testing across a variety of browsers.
